server/src/main/java/io/crate/execution/ddl/tables/TransportAddColumnAction.java

```diff
@@ -0,0 +1,181 @@
+            Map<String, Object> indexMapping = indexMetadata.mapping().sourceAsMap();
+
+            mergeDeltaIntoExistingMapping(indexMapping, request);
+
+            IndexMetadata.Builder imBuilder = IndexMetadata.builder(indexMetadata);
+            var updatedMetadata = new MappingMetadata(Constants.DEFAULT_MAPPING_TYPE, indexMapping); // throws IOException
+            imBuilder.putMapping(updatedMetadata).mappingVersion(1 + imBuilder.mappingVersion()); // Similar to MetadataMappingService
+            metadataBuilder.put(imBuilder); // increments version.
+        }
+
+        return ClusterState.builder(currentState).metadata(metadataBuilder).build();
+    }
+
+    private static void mergeDeltaIntoExistingMapping(Map<String, Object> existingMapping, AddColumnRequest request) {
+        // _meta fields
+        if (request.isPrimaryKey()) {
+            Maps.mergeInto(existingMapping, "_meta", List.of("primary_keys"), request.columnRef().column().fqn(),
+                (map, key, value) -> ((ArrayList) map.computeIfAbsent(key, k -> new ArrayList<>())).add(value)
+            );
+        }
+        if (request.columnRef().isNullable() == false) {
+            Maps.mergeInto(existingMapping, "_meta", List.of("constraints", "not_null"), request.columnRef().column().fqn(),
+                (map, key, value) -> ((ArrayList) map.computeIfAbsent(key, k -> new ArrayList<>())).add(value)
+            );
+        }
+        if (request.generatedExpression() != null) {
+            Maps.mergeInto(existingMapping, "_meta", List.of("generated_columns"), Map.entry(request.columnRef().column().fqn(), request.generatedExpression()),
+                (map, key, value) -> {
+                    Map.Entry<String, String> nameAndExpression = (Map.Entry) value;
+                    ((LinkedHashMap) map.computeIfAbsent(key, k -> new LinkedHashMap<>())).put(nameAndExpression.getKey(), nameAndExpression.getValue());
+                }
+            );
+        }
+        for (AddColumnRequest.StreamableCheckConstraint checkConstraint: request.checkConstraints()) {
+            Maps.mergeInto(existingMapping, "_meta", List.of("check_constraints"), checkConstraint,
+                (map, key, value) -> {
+                    AddColumnRequest.StreamableCheckConstraint constraint = (AddColumnRequest.StreamableCheckConstraint) value;
+                    ((LinkedHashMap) map.computeIfAbsent(key, k -> new LinkedHashMap<>())).put(constraint.name(), constraint.expression());
+                }
+            );
+        }
+
+        // properties field (columnstore, index flags and column type specific properties)
+        Maps.mergeInto(existingMapping, "properties", List.of(), request.columnProperties(),
+            (map, key, value) -> {
+                HashMap<String, Object> colProps = (HashMap) value;
+                ((LinkedHashMap) map.computeIfAbsent(key, k -> new LinkedHashMap<>())).put(request.columnRef().column().fqn(), colProps);
+            }
```

Comment:
Choose of data structures was a bit confusing to me - I aligned with 
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/analyze/AnalyzedTableElements.java#L112-L149 but then faced casting issues like `class java.util.ArrayList cannot be cast to class java.util.LinkedHashSet` - I checked in runtime and aligned with runtime values. Apparently should have aligned with DocIndexMetadata, which I did now.